src/backend/distributed/executor/local_executor.c

```diff
@@ -0,0 +1,579 @@
+		{
+			/*
+			 * At this point, the task has a single placement (e.g,. anchor shard  is
+			 * distributed table's shard). So, it is either added to local or remote
+			 * taskList.
+			 */
+			if (localTaskPlacementList == NIL)
+			{
+				*remoteTaskList = lappend(*remoteTaskList, task);
+			}
+			else
+			{
+				*localTaskList = lappend(*localTaskList, task);
+			}
+		}
+		else
+		{
+			Task *localTask = copyObject(task);
+			Task *remoteTask = NULL;
+
+			/*
+			 * At this point, we're dealing with reference tables or intermediate results
+			 * where the task has placements on both local and remote nodes. We always
+			 * prefer to use local placement, and require remote placements only for
+			 * modifications.
+			 */
+
+			localTask->taskPlacementList = localTaskPlacementList;
+			*localTaskList = lappend(*localTaskList, localTask);
+
+			if (readOnly)
+			{
+				/* read-only tasks should only be executed on the local machine */
+			}
+			else
+			{
+				remoteTask = copyObject(task);
+				remoteTask->taskPlacementList = remoteTaskPlacementList;
+
+				*remoteTaskList = lappend(*remoteTaskList, remoteTask);
+			}
+		}
+	}
+}
+
+
+/*
+ * SplitLocalAndRemotePlacements is a helper function which iterates over the input
+ * taskPlacementList and puts the placements into corresponding list of either
+ * localTaskPlacementList or remoteTaskPlacementList.
+ */
+static void
+SplitLocalAndRemotePlacements(List *taskPlacementList, List **localTaskPlacementList,
+							  List **remoteTaskPlacementList)
+{
+	ListCell *placementCell = NULL;
+	int32 localGroupId = GetLocalGroupId();
+
+	*localTaskPlacementList = NIL;
+	*remoteTaskPlacementList = NIL;
+
+	foreach(placementCell, taskPlacementList)
+	{
+		ShardPlacement *taskPlacement =
+			(ShardPlacement *) lfirst(placementCell);
+
+		if (taskPlacement->groupId == localGroupId)
+		{
+			*localTaskPlacementList = lappend(*localTaskPlacementList, taskPlacement);
+		}
+		else
+		{
+			*remoteTaskPlacementList = lappend(*remoteTaskPlacementList, taskPlacement);
+		}
+	}
+}
+
+
+/*
+ * ExecuteLocalTaskList goes over the task list and executes them locally.
+ * The returning tuples (if any) is stored in the CitusScanState.
+ *
+ * The function returns totalRowsProcessed.
+ */
+static uint64
+ExecuteLocalTaskList(CitusScanState *node, List *taskList)
+{
+	EState *executorState = ScanStateGetExecutorState(node);
+	ParamListInfo paramListInfo = copyParamList(executorState->es_param_list_info);
+	Oid *paramOids = GetParamListOids(paramListInfo);
+	int numParams = paramListInfo != NULL ? paramListInfo->numParams : 0;
+
+	ListCell *taskCell = NULL;
+
+	uint64 totalRowsProcessed = 0;
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+
+		PlannedStmt *localPlan = NULL;
+		int cursorOptions = 0;
+		const char *shardQueryString = task->queryString;
+		Query *shardQuery = ParseQueryString(shardQueryString, paramOids, numParams);
```

Comment:
Note: `ParseQueryString ()` performs slightly (~15-25%) better than this [code-block](https://github.com/citusdata/citus/pull/2830/files#diff-aff2f55a09f9f9619f28774e6fb18965R363-R378) and seems easier to read. Let me know if you think otherwise.

Reply:
I guess it makes sense for simple queries, which is the common case that we're trying to optimise. We could probably do even better if we replaced table OIDs directly with shard OIDs. (not necessary)